Demonstrate global energy conservation for INS FE

lindsayad opened the pull request:
Closes #22074

lindsayad committed:
Demonstrate global energy conservation for INS FE

Closes #22074

GiudGiud commented on modules/navier_stokes/test/tests/finite_element/ins/energy-conservation/tests:
```diff
@@ -0,0 +1,19 @@
+  [conservation]
+    requirement = 'The system shall exhibit global conservation of energy when using the continuous-Galerkin finite element spatial discretization with streamline-upwind Petrov-Galerkin stabilization and with'
+    [q2q1]
+      type = Exodiff
```
exodiff is overkill, csvdiff is more appropriate and will cost less memory

GiudGiud replied:
jus trying to be consistent with what I m requesting from the reactor module PR makers

lindsayad replied:
:+1:

GiudGiud commented on modules/navier_stokes/test/tests/finite_element/ins/energy-conservation/q2q1.i:
```diff
@@ -0,0 +1,179 @@
+    function = 'x + y'
+    block = 0
+  []
+  [sum]
```
```suggestion
  [energy_balance]
```

GiudGiud commented on modules/navier_stokes/test/tests/finite_element/ins/energy-conservation/q2q1.i:
```diff
@@ -0,0 +1,179 @@
+    block = 0
+    function = 'x + y'
+  []
+  [remove_null_space]
+    variable = T
+    block = 0
+    type = Reaction
+  []
```
you prefer this to setting a flux BC on temperature? this option would be more physical

lindsayad replied:
A flux BC on temperature does not remove the null space as the boundary flux terms only involve gradients of temperature (e.g. conduction). The advection term directly applies the incompressibility constraint and consequently does not perform integration by parts. This could be something we change in the future though

GiudGiud replied:
with an initial condition the evolution of temperature should be well defined.

GiudGiud replied:
like physically. numerically I can see your argument that there is nothing "holding" temperature only its gradient, that there is a null space

lindsayad replied:
I did end up going with an integrated BC, a convective heat flux BC on the right boundary. Nice suggestion for physicality!

GiudGiud replied:
oh my bad. This is a steady problem so the initial condition isnt necessarily relevant

GiudGiud commented on modules/navier_stokes/test/tests/finite_element/ins/energy-conservation/q1q1.i:
```diff
@@ -0,0 +1,180 @@
+    function = 'x + y'
+    block = 0
+  []
+  [sum]
```
```suggestion
  [energy_balance]
```

GiudGiud commented on modules/navier_stokes/doc/content/source/postprocessors/INSElementIntegralEnergyAdvection.md:
```diff
@@ -0,0 +1,40 @@
+`INSElementIntegralEnergyAdvection`. The former retrieves the density and
+specific heat capacity as `ADReal` material properties while the former
+retrieves the same properties as `Real`.
+
```
you should add a note to say why this is more appropriate than doing surface based balances

lindsayad replied:
:+1:

GiudGiud reviewed (changes requested):
interesting.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1169226/) on 614e32a wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/22075/site)

This comment will be updated on new commits.

lindsayad committed:
Use convective heat flux bc to remove nullspace

Also convert to CSVDiff as opposed to exodiff since our purpose
here is to demonstrate conservation in the postprocessors

lindsayad committed:
sum -> energy_balance for pp name

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

lindsayad committed:
Describe origin of energy advection equation form

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

GiudGiud commented on modules/navier_stokes/doc/content/source/postprocessors/INSElementIntegralEnergyAdvection.md:
```diff
@@ -0,0 +1,46 @@
+
+!syntax description /Postprocessors/INSElementIntegralEnergyAdvection
+
+This class performs global integration of the energy advection term defined in
```
```suggestion
This class performs volumetric (global, or by block) integration of the energy advection term defined in
```

GiudGiud commented on modules/navier_stokes/src/postprocessors/INSElementIntegralEnergyAdvection.C:
```diff
@@ -0,0 +1,52 @@
+INSElementIntegralEnergyAdvectionTempl<is_ad>::validParams()
+{
+  InputParameters params = ElementIntegralPostprocessor::validParams();
+  params.addClassDescription("Computes the volume integrated energy transport due to advection");
```
this is not the "transport" that s not a quantity (even though I understand what you mean)

This is more like the contribution of advection to the energy balance. 

Computes the net volumetric balance of energy transported by advection ??

lindsayad replied:
I think that sounds reasonable

GiudGiud commented on modules/navier_stokes/test/tests/finite_element/ins/energy-conservation/q1q1.i:
```diff
@@ -0,0 +1,184 @@
+  [convective_heat_transfer]
+    type = ConvectiveHeatFluxBC
+    variable = T
+    T_infinity = 0
```
oh yeah this should remove the null space?

GiudGiud replied:
even though it s a neumann BC right

lindsayad replied:
yessir
